pydantic/json_schema.py

```diff
@@ -1043,17 +1043,30 @@ def default_schema(self, schema: core_schema.WithDefaultSchema) -> JsonSchemaVal
         #     return json_schema
 
         # we reflect the application of custom plain, no-info serializers to defaults for
-        # json schemas viewed in serialization mode
+        # JSON Schemas viewed in serialization mode:
         # TODO: improvements along with https://github.com/pydantic/pydantic/issues/8208
-        # TODO: improve type safety here
-        if self.mode == 'serialization':
-            if (
-                (ser_schema := schema['schema'].get('serialization', {}))
-                and (ser_func := ser_schema.get('function'))
-                and ser_schema.get('type') == 'function-plain'  # type: ignore
-                and ser_schema.get('info_arg') is False  # type: ignore
-            ):
+        if (
+            self.mode == 'serialization'
+            and (ser_schema := schema['schema'].get('serialization'))
+            and (ser_func := ser_schema.get('function'))
+            and ser_schema.get('type') == 'function-plain'
+            and not ser_schema.get('info_arg')
+            and not (default is None and ser_schema.get('when_used') == 'json-unless-none')
+        ):
+            try:
                 default = ser_func(default)  # type: ignore
+            except Exception:
+                # It might be that the provided default needs to be validated first
+                # (assuming `validate_default` is enabled). However, we can't perform
+                # such validation during JSON Schema generation so we don't support
+                # this pattern for now.
+                # (One example is when using `foo: ByteSize = '1MB'`, which validates and
+                # serializes as an int. In this case, `ser_func` is `int` and `int('1MB')` fails).
+                self.emit_warning(
+                    'non-serializable-default',
+                    f'Unable to serialize value {default!r} with the plain serializer; excluding default from JSON schema',
+                )
+                return json_schema
```

Comment:
Could we do something like this?

https://github.com/pydantic/pydantic/blob/0c97049bd40ac95cce7bbebad8369b9426af4b09/pydantic/json_schema.py#L2017-L2044

I feel like we shouldn't be skipping that logic here, we could maybe call that function?

Reply:
I don't think it makes sense to do so, or we may end up with a wrong default value in the JSON Schema (in the related test fixed, `'1MB'` was set as the default value in the JSON Schema, but this isn't true as the default is `1000000`). In my opinion, it's better to have no value than a wrong one.

However, we emit a warning here, so I'm ok to call `encode_default`.

Reply:
That's fair - I wonder if we could use a similar pattern though + attempt to validate the default + include that in the JSON schema. Feels messy to do validation here, I admit.

I'm not opposed to the warning you have in place, I think that's good!

Reply:
> That's fair - I wonder if we could use a similar pattern though + attempt to validate the default + include that in the JSON schema. Feels messy to do validation here, I admit.

I think we can do this in a separate PR, if someone asks for it. I'm not keen to go ahead and do this now.

Change looks good to me, thanks for including the helpful warning.